### PR TITLE
Add Neon optimised RGB2Lab conversion

### PR DESCRIPTION
**Pull request by jondea**
A Neon specific implementation of RGB2Lab increases single threaded performance by ~25%, here's the numbers run on aws c6gd.4xlarge with gcc 9.3 (numbers are similar using gcc 10)

| Test set   | Test number |    After/before ratio | Speedup with 1/million bounds [%] |
|------------|-------------|----------|-----------------------------------|
| cvtColor8u |           8 |  0.76835 |                        23.2 ± 0.2 |
| cvtColor8u |          34 |  0.76204 |                        23.8 ± 0.2 |
| cvtColor8u |          67 |  0.76667 |                        23.3 ± 0.2 |
| cvtColor8u |          69 |  0.76773 |                        23.2 ± 0.2 |
| cvtColor8u |          71 |  0.76231 |                        23.8 ± 0.2 |
| cvtColor8u |          73 |  0.76184 |                        23.8 ± 0.2 |
| cvtColor8u |          90 |  0.76851 |                        23.1 ± 0.2 |
| cvtColor8u |         103 |  0.76143 |                        23.9 ± 0.2 |
| cvtColor8u |         128 |  0.73870 |                        26.1 ± 0.1 |
| cvtColor8u |         154 |  0.73760 |                        26.2 ± 0.2 |
| cvtColor8u |         187 |  0.73891 |                        26.1 ± 0.1 |
| cvtColor8u |         189 |  0.73889 |                        26.1 ± 0.1 |
| cvtColor8u |         191 |  0.73802 |                        26.2 ± 0.2 |
| cvtColor8u |         193 |  0.73817 |                        26.2 ± 0.2 |
| cvtColor8u |         210 |  0.73879 |                        26.1 ± 0.1 |
| cvtColor8u |         223 |  0.73745 |                        26.3 ± 0.2 |
| cvtColor8u |         248 |  0.73756 |                        26.2 ± 0.1 |
| cvtColor8u |         274 |  0.73613 |                        26.4 ± 0.1 |
| cvtColor8u |         307 |  0.73768 |                        26.2 ± 0.1 |
| cvtColor8u |         309 |  0.73767 |                        26.2 ± 0.1 |
| cvtColor8u |         311 |  0.73676 |                        26.3 ± 0.2 |
| cvtColor8u |         313 |  0.73672 |                        26.3 ± 0.2 |
| cvtColor8u |         330 |  0.73748 |                        26.3 ± 0.1 |
| cvtColor8u |         343 |  0.73591 |                        26.4 ± 0.1 |

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,docs,ARMv8,ARMv7
```

### CODE REVIEW

**Comment by alalek** [comment]
Marking this **RFC**, because it doesn't follow OpenCV guidelines to avoid using of raw native intrinsics in OpenCV modules.

**Comment by jondea** [comment]
Hi @alalek, thank you for looking into this. We (me and @fpetrogalli) submitted it like this because we weren't sure what correct approach was. Also, sorry if this is an silly question, but what does it mean to mark it as RFC?

One possible solution to the raw intrinsics is to keep the `#if CV_NEON` block in color_lab.cpp but rewrite it using the HAL intrinsics. Another solution would be to split it into a neon specific file, like in the case of `resize.cpp`, `resize.avx2.cpp` and `resize.sse4_1.cpp` for example. Are either of these acceptable or preferable? Or is there a another way which would achieve the same goal?

**Comment by vpisarev** [comment]
@jondea, thank you for the contribution!
as @alalek said, for the tiny OpenCV core team it's simply unfeasible to maintain separate code branches for the growing amount of code and the growing number of platforms that we support. With time we hope to port most of the remaining native branches to HAL/universal intrinsics. There will be some exceptions, like deep learning, where the amount of critical kernels is not that big and where we can afford separate branches, but overall universal intrinsics is the preferable (by far) option.

I'd start with the first option that you suggested - keep the separate branch under CV_NEON, but rewrite it using HAL intrinsics. I briefly looked at the current implementation and I found it too bulky for the equivalent C code that it accelerates. So, I'm 60-80% sure that the HAL code that you write will be faster than the existing implementation not just on ARM, but on the other platforms as well. And then we will just replace that code with yours, i.e. remove `#if CV_NEON ... #endif` around your code and remove the other branch.

**Comment by jondea** [comment]
The changes have been rewritten to use just HAL intrinsics, any feedback would be appreciated.

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [inline modules/imgproc/src/color_lab.cpp:1596]: I think you might want to use `v_uint8::nlanes` here instead of 16?

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [inline modules/imgproc/src/color_lab.cpp:1655]: Nit: one empty line should make it.

**Review by fpetrogalli — COMMENTED** [review]
Hi @jondea,

just a couple of minor observations.

Thank you for your work.

Francesco

File: modules/imgproc/src/color_lab.cpp
Comment (jondea) [reply modules/imgproc/src/color_lab.cpp:1655]: Done

File: modules/imgproc/src/color_lab.cpp
Comment (jondea) [reply modules/imgproc/src/color_lab.cpp:1670]: Done

File: modules/imgproc/src/color_lab.cpp
Comment (jondea) [reply modules/imgproc/src/color_lab.cpp:1596]: Done, also changed the types throughout to not hard code the number of lanes

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [inline modules/imgproc/src/color_lab.cpp:1677]:
```suggestion
#endif // CV_NEON
```

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [inline modules/imgproc/src/color_lab.cpp:1613]: May I suggest to use a static (inline?) function here instead of a macro? Should make the code easier to read. It will likely be inlined so the final code will perform as the one produced with the macro.

You should also guard it with `CV_NEON` to prevent it being built when not used.

Comment (jondea) [reply modules/imgproc/src/color_lab.cpp:1613]: Done, although it was neater to do in a member function of RGB2Lab_b

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [inline modules/imgproc/src/color_lab.cpp:1626]:
```suggestion
        vL = v_fma(vfY, v_setall_s32(Lscale), v_setall_s32(Lshift+labDescaleShift));
```

**Review by fpetrogalli — APPROVED** [review]
LGTM, with a nit.

Final word on the maintainers, of course!

Thank you, Francesco

**Comment by jondea** [comment]
@vpisarev @alalek is there anything else which needs to be done before this can be merged?

**Comment by jondea** [comment]
Hi @vpisarev @alalek would you be able to take another look at this please and let me know if it can be merged?

**Comment by vpisarev** [comment]
@jondea, thank you very much! I tested the code both on Mac-Intel and Mac-ARM (M1), it works well, the claimed acceleration is achieved. On Intel it's no slower than the previous version, but, unfortunately, it's 128-bit only.

In any case, it can be merged as-is, and later we can modify this code to use some new variations of `v_lut()` intrinsic.

:+1:

**Review by vpisarev — APPROVED** [review]

File: modules/imgproc/src/color_lab.cpp
Comment (asmorkalov) [inline modules/imgproc/src/color_lab.cpp:1576]: It should be `#if CV_SIMD`.

File: modules/imgproc/src/color_lab.cpp
Comment (asmorkalov) [inline modules/imgproc/src/color_lab.cpp:1653]: The same.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/imgproc/src/color_lab.cpp
Comment (fpetrogalli) [reply modules/imgproc/src/color_lab.cpp:1576]: Hi @asmorkalov  - this code is specific to 128-bit. If we replace `CV_NEON` with `CV_SIMD`, the version supporting wider vectors at line 1692 will stop running.

Would replacing `CV_NEON` with `CV_SIMD128` a better option?

**Comment by fpetrogalli** [comment]
> @jondea, thank you very much! I tested the code both on Mac-Intel and Mac-ARM (M1), it works well, the claimed acceleration is achieved. On Intel it's no slower than the previous version, but, unfortunately, it's 128-bit only.

@vpisarev , @jondea  is working on an equivalent version that uses SVE2 intrinsics. He is using the intrinsic `svld1uh_gather_s32index_s32` for the variation of `v_lut()` that does a gather from the indexes. It is a Vector Length Agnositc (VLA) version, so it could be ported easily to HAL once we make the `nlanes` field to be a runtime value and we have the correspondent indexed `lut` intrinsic.

File: modules/imgproc/src/color_lab.cpp
Comment (alalek) [reply modules/imgproc/src/color_lab.cpp:1576]: @asmorkalov `CV_SIMD` is below. This is dedicated branch for NEON only (existed `v_lut()`-based approach is not very fast with NEON).

**Pull request merged**
